builtin/providers/aws/resource_aws_instance.go

```diff
@@ -514,44 +530,130 @@ func resourceAwsInstanceCreate(d *schema.ResourceData, meta interface{}) error {
 		runOpts.BlockDeviceMappings = blockDevices
 	}
 
-	// Create the instance
-	log.Printf("[DEBUG] Run configuration: %#v", runOpts)
-	var err error
-
-	var runResp *ec2.Reservation
-	for i := 0; i < 5; i++ {
-		runResp, err = conn.RunInstances(runOpts)
-		if awsErr, ok := err.(awserr.Error); ok {
-			// IAM profiles can take ~10 seconds to propagate in AWS:
-			//  http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#launch-instance-with-role-console
-			if awsErr.Code() == "InvalidParameterValue" && strings.Contains(awsErr.Message(), "Invalid IAM Instance Profile") {
-				log.Printf("[DEBUG] Invalid IAM Instance Profile referenced, retrying...")
-				time.Sleep(2 * time.Second)
-				continue
-			}
+	instanceID := ""
+	onDemand := true
+
+	if sp := d.Get("spot_price").(string); sp != "" {
+		// FIXME: check that instance type is not t2.micro, t2.small, or t2.medium
+
+		onDemand = false
+		// FIXME: validate spot price casts to a sensible decimal
+
+		spotPlacement := &ec2.SpotPlacement{
+			AvailabilityZone: aws.String(d.Get("availability_zone").(string)),
+			GroupName:        aws.String(d.Get("placement_group").(string)),
+		}
+
+		launchSpec := &ec2.RequestSpotLaunchSpecification{
+			Placement:           spotPlacement,
+			BlockDeviceMappings: runOpts.BlockDeviceMappings,
+			ImageID:             runOpts.ImageID,
+			InstanceType:        runOpts.InstanceType,
+			UserData:            runOpts.UserData,
+			EBSOptimized:        runOpts.EBSOptimized,
+			IAMInstanceProfile:  runOpts.IAMInstanceProfile,
+			NetworkInterfaces:   runOpts.NetworkInterfaces,
+		}
+
+		if runOpts.SubnetID != nil {
+			launchSpec.SubnetID = runOpts.SubnetID
+		}
+
+		if runOpts.SecurityGroupIDs != nil {
+			launchSpec.SecurityGroupIDs = runOpts.SecurityGroupIDs
+		}
+
+		if runOpts.SecurityGroups != nil {
+			launchSpec.SecurityGroups = runOpts.SecurityGroups
+		}
+
+		if runOpts.KeyName != nil {
+			launchSpec.KeyName = runOpts.KeyName
+		}
+
+		// Build the spot instance struct
+		spotOpts := &ec2.RequestSpotInstancesInput{
+			SpotPrice:           aws.String(sp), // Required
+			InstanceCount:       aws.Long(1),
+			LaunchSpecification: launchSpec,
+		}
+
+		// Make the spot instance request
+		var spotResp *ec2.RequestSpotInstancesOutput
+		spotResp, err := conn.RequestSpotInstances(spotOpts)
+
+		request_id := *spotResp.SpotInstanceRequests[0].SpotInstanceRequestID
+		d.Set("spot_request_id", request_id)
+
+		spotStateConf := &resource.StateChangeConf{
+			// http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-bid-status.html
+			Pending:    []string{"start", "pending-evaluation", "pending-fulfillment"},
+			Target:     "fulfilled",
+			Refresh:    SpotInstanceStateRefreshFunc(conn, request_id),
+			Timeout:    10 * time.Minute,
```

Comment:
This all comes down to how fast amazon resolves your bid. Typically it's about 5 minutes or less, but I've seen it take as long as half an hour in rare cases.

Making it configurable wouldn't hurt - to be honest though this is just copy pasta. If someone decides they want a spot instance, they should be aware that it will take some time for their bid to resolve